### PR TITLE
feat: implement secrets

### PR DESCRIPTION
**Pull request by WiredSound**
I spent a considerable amount of time trying out different ways of doing this but ultimately settle on a pretty simple system wherein `PgPool` implements a `SecretStore` which fetches secrets stored in a table appropriately named `secrets` in the database. Storing secrets directly in the user-accessible database allows for them to simply set secrets by just connecting to the database directly. Secrets are also read from a `Secrets.toml` file at time of deployment.

### CODE REVIEW

File: service/src/secrets.rs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    for<'c> usize: ColumnIndex<<DB as Database>::Row>,
+    for<'c> Query<'c, Postgres, PgArguments>: Execute<'c, DB>
+{
+    // TODO: Don't restrict to Postgres types above.
+
+    const GET_QUERY: &'static str;
+    const SET_QUERY: &'static str;
+    const CREATE_TABLE_QUERY: &'static str;
+
+    /// Read the secret with the given key from the database. Will an error if a secret with the
```
Comment (chesedo) [inline service/src/secrets.rs:50]:
```suggestion
    /// Read the secret with the given key from the database. Will error if a secret with the
```

File: service/src/lib.rs [hunk 2]
```diff
@@ -165,6 +165,11 @@ use tokio::runtime::Runtime;
 pub mod error;
 pub use error::Error;
 
+#[cfg(feature = "sqlx-postgres")]
+pub mod secrets;
+#[cfg(feature = "sqlx-postgres")]
```
Comment (chesedo) [inline service/src/lib.rs:170]: Should this maybe be behind a secrets feature gate?

File: api/Cargo.toml [hunk 3]
```diff
@@ -29,4 +29,4 @@ base64 = "0.13.0"
 rand = "0.8.5"
 
 shuttle-common = { version = "0.2.6", path = "../common" }
-shuttle-service = { version = "0.2.6", path = "../service", features = [ "loader" ] }
+shuttle-service = { version = "0.2.6", path = "../service", features = [ "loader", "sqlx-postgres" ] }
```
Comment (chesedo) [inline api/Cargo.toml:32]: Is it possible to not turn the postgres feature on? Maybe the secret store should be in common instead?

File: api/src/args.rs [hunk 4]
```diff
@@ -30,7 +30,7 @@ pub struct Args {
     )]
     pub(crate) bind_addr: IpAddr,
     #[structopt(long, about = "Fully qualified domain name deployed services are reachable at", parse(try_from_str = parse_fqdn))]
-    pub(crate) proxy_fqdn: FQDN,
+    pub(crate) proxy_fqdn: FQDN
```
Comment (chesedo) [inline api/src/args.rs:33]: For some reason the rustfmt file seems to not be respected by your formater :joy:

File: cargo-shuttle/src/client.rs [hunk 5]
```diff
@@ -138,6 +157,33 @@ pub(crate) async fn deploy(
...
 
+pub(crate) async fn secrets(
+    api_url: ApiUrl,
+    api_key: &ApiKey,
+    project: &ProjectName,
+    secrets: HashMap<String, String>
+) -> Result<()> {
+    if secrets.is_empty() {
+        Ok(())
+    }
```
Comment (chesedo) [inline cargo-shuttle/src/client.rs:168]:
```suggestion
    if secrets.is_empty() {
        return Ok(());
    }
```
This is called a guard clause and will allow you to remove the nesting created by the `else` block (to be removed manually :sweat_smile: )

**Review by chesedo — CHANGES_REQUESTED** [review]
Looking good!! Left some minor suggestions

File: api/Cargo.toml (hunk 3 above)
Comment (WiredSound) [reply api/Cargo.toml]: But then that would surely just move the Postgres dependency into common instead - would that still be advantageous?

**Review by chesedo — APPROVED** [review]

**Pull request merged**